# Assignment week 5 - Gender recognition of voice

I chose the dataset of gender recognition, simply because I wasn't excited about the other options so I wanted to see if I could predict the gender based upon both voice and speech. The dataset has 3168 entries and as they explain: "The voice samples are pre-processed by acoustic analysis in R using the seewave and tuneR packages".

In [2]:
import seaborn as sns
import sklearn as sk
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split

In [3]:
df = pd.read_csv('voice.csv')
df.head(15)

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0   0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1   0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2   0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3   0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4   0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5   0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6   0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7   0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8   0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9   0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   
10  0.157021  0.071943  0.168160  0.101430  0.216740  0.115310   0.979442   
11  0.138551  0.077054  0.127527  0.087314  0.202739  0.115426   1.626770   
12  0.137343  0.080877  0.124263  0.083145  0.209227  0.126082   1.378728   
13  0.181225  0.060042  0.190953  0.128839  0.229532  0.100693   1.369430   
14  0.183115  0.066982  0.191233  0.129149  0.240152  0.111004   3.568104   

           kurt    sp.ent       sfm  ...  centroid   meanfun    minfun  \
0    274.402906  0.893369  0.491918  ...  0.059781  0.084279  0.015702   
1    634.613855  0.892193  0.513724  ...  0.066009  0.107937  0.015826   
2   1024.927705  0.846389  0.478905  ...  0.077316  0.098706  0.015656   
3      4.177296  0.963322  0.727232  ...  0.151228  0.088965  0.017798   
4      4.333713  0.971955  0.783568  ...  0.135120  0.106398  0.016931   
5      8.308895  0.963181  0.738307  ...  0.132786  0.110132  0.017112   
6      5.987498  0.967573  0.762638  ...  0.150762  0.105945  0.026230   
7      4.766611  0.959255  0.719858  ...  0.160514  0.093052  0.017758   
8      4.070284  0.970723  0.770992  ...  0.142239  0.096729  0.017957   
9      4.787310  0.975246  0.804505  ...  0.134329  0.105881  0.019300   
10     3.974223  0.965249  0.733693  ...  0.157021  0.088894  0.022069   
11     6.291365  0.966004  0.752042  ...  0.138551  0.104199  0.019139   
12     5.008952  0.963514  0.736150  ...  0.137343  0.092644  0.016789   
13     5.475600  0.937446  0.537080  ...  0.181225  0.131504  0.025000   
14    35.384748  0.940333  0.571394  ...  0.183115  0.102799  0.020833   

      maxfun   meandom    mindom    maxdom   dfrange   modindx  label  
0   0.275862  0.007812  0.007812  0.007812  0.000000  0.000000   male  
1   0.250000  0.009014  0.007812  0.054688  0.046875  0.052632   male  
2   0.271186  0.007990  0.007812  0.015625  0.007812  0.046512   male  
3   0.250000  0.201497  0.007812  0.562500  0.554688  0.247119   male  
4   0.266667  0.712812  0.007812  5.484375  5.476562  0.208274   male  
5   0.253968  0.298222  0.007812  2.726562  2.718750  0.125160   male  
6   0.266667  0.479620  0.007812  5.312500  5.304688  0.123992   male  
7   0.144144  0.301339  0.007812  0.539062  0.531250  0.283937   male  
8   0.250000  0.336476  0.007812  2.164062  2.156250  0.148272   male  
9   0.262295  0.340365  0.015625  4.695312  4.679688  0.089920   male  
10  0.117647  0.460227  0.007812  2.812500  2.804688  0.200000   male  
11  0.262295  0.246094  0.007812  2.718750  2.710938  0.132351   male  
12  0.213333  0.481671  0.015625  5.015625  5.000000  0.088500   male  
13  0.275862  1.277114  0.007812  2.804688  2.796875  0.416550   male  
14  0.275862  1.245739  0.203125  6.742188  6.539062  0.139332   male  

[15 rows x 21 columns]

By the use of a correlation matrix we can determine which variables to use for the actual prediction. Let's have a look!

In [4]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   

              skew      kurt    sp.ent       sfm      mode  centroid  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
sd        0.314597  0.346241  0.716620  0.838086 -0.529150 -0.739039   
median   -0.257407 -0.243382 -0.502005 -0.661690  0.677433  0.925445   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  0.591277  0.911416   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  0.486857  0.740997   
IQR       0.249497  0.316185  0.640813  0.663601 -0.403764 -0.627605   
skew      1.000000  0.977020 -0.195459  0.079694 -0.434859 -0.322327   
kurt      0.977020  1.000000 -0.127644  0.109884 -0.406722 -0.316036   
sp.ent   -0.195459 -0.127644  1.000000  0.866411 -0.325298 -0.601203   
sfm       0.079694  0.109884  0.866411  1.000000 -0.485913 -0.784332   
mode     -0.434859 -0.406722 -0.325298 -0.485913  1.000000  0.687715   
centroid -0.322327 -0.316036 -0.601203 -0.784332  0.687715  1.000000   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  0.324771  0.460844   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  0.385467  0.383937   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  0.172329  0.274004   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  0.491479  0.536666   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  0.198150  0.229261   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  0.477187  0.519528   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  0.473775  0.515570   
modindx  -0.169325 -0.205539  0.198074  0.211477 -0.182344 -0.216979   

           meanfun    minfun    maxfun   meandom    mindom    maxdom  \
meanfreq  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
sd       -0.466281 -0.345609 -0.129662 -0.482726 -0.357667 -0.482278   
median    0.414909  0.337602  0.251328  0.455943  0.191169  0.438919   
Q25       0.545035  0.320994  0.199841  0.467403  0.302255  0.459683   
Q75       0.155091  0.258002  0.285584  0.359181 -0.023750  0.335114   
IQR      -0.534462 -0.222680 -0.069588 -0.333362 -0.357037 -0.337877   
skew     -0.167668 -0.216954 -0.080861 -0.336848 -0.061608 -0.305651   
kurt     -0.194560 -0.203201 -0.045667 -0.303234 -0.103313 -0.274500   
sp.ent   -0.513194 -0.305826 -0.120738 -0.293562 -0.294869 -0.324253   
sfm      -0.421066 -0.362100 -0.192369 -0.428442 -0.289593 -0.436649   
mode      0.324771  0.385467  0.172329  0.491479  0.198150  0.477187   
centroid  0.460844  0.383937  0.274004  0.536666  0.229261  0.519528   
meanfun   1.000000  0.339387  0.311950  0.270840

We need the variable label for our prediction and it doesn't show up in te correlation. One way of fixing this problem is by making this variable into dummie variables, this way it will be numerical (int).

In [5]:
#make the dummies out of the variable label
dummies = pd.get_dummies(df['label'])
dummies

#and now let's add them to the columns!
df = pd.concat([df, dummies], axis=1)
df.head(12)

meanfreq        sd    median       Q25       Q75       IQR       skew  \
0   0.059781  0.064241  0.032027  0.015071  0.090193  0.075122  12.863462   
1   0.066009  0.067310  0.040229  0.019414  0.092666  0.073252  22.423285   
2   0.077316  0.083829  0.036718  0.008701  0.131908  0.123207  30.757155   
3   0.151228  0.072111  0.158011  0.096582  0.207955  0.111374   1.232831   
4   0.135120  0.079146  0.124656  0.078720  0.206045  0.127325   1.101174   
5   0.132786  0.079557  0.119090  0.067958  0.209592  0.141634   1.932562   
6   0.150762  0.074463  0.160106  0.092899  0.205718  0.112819   1.530643   
7   0.160514  0.076767  0.144337  0.110532  0.231962  0.121430   1.397156   
8   0.142239  0.078018  0.138587  0.088206  0.208587  0.120381   1.099746   
9   0.134329  0.080350  0.121451  0.075580  0.201957  0.126377   1.190368   
10  0.157021  0.071943  0.168160  0.101430  0.216740  0.115310   0.979442   
11  0.138551  0.077054  0.127527  0.087314  0.202739  0.115426   1.626770   

           kurt    sp.ent       sfm  ...    minfun    maxfun   meandom  \
0    274.402906  0.893369  0.491918  ...  0.015702  0.275862  0.007812   
1    634.613855  0.892193  0.513724  ...  0.015826  0.250000  0.009014   
2   1024.927705  0.846389  0.478905  ...  0.015656  0.271186  0.007990   
3      4.177296  0.963322  0.727232  ...  0.017798  0.250000  0.201497   
4      4.333713  0.971955  0.783568  ...  0.016931  0.266667  0.712812   
5      8.308895  0.963181  0.738307  ...  0.017112  0.253968  0.298222   
6      5.987498  0.967573  0.762638  ...  0.026230  0.266667  0.479620   
7      4.766611  0.959255  0.719858  ...  0.017758  0.144144  0.301339   
8      4.070284  0.970723  0.770992  ...  0.017957  0.250000  0.336476   
9      4.787310  0.975246  0.804505  ...  0.019300  0.262295  0.340365   
10     3.974223  0.965249  0.733693  ...  0.022069  0.117647  0.460227   
11     6.291365  0.966004  0.752042  ...  0.019139  0.262295  0.246094   

      mindom    maxdom   dfrange   modindx  label  female  male  
0   0.007812  0.007812  0.000000  0.000000   male       0     1  
1   0.007812  0.054688  0.046875  0.052632   male       0     1  
2   0.007812  0.015625  0.007812  0.046512   male       0     1  
3   0.007812  0.562500  0.554688  0.247119   male       0     1  
4   0.007812  5.484375  5.476562  0.208274   male       0     1  
5   0.007812  2.726562  2.718750  0.125160   male       0     1  
6   0.007812  5.312500  5.304688  0.123992   male       0     1  
7   0.007812  0.539062  0.531250  0.283937   male       0     1  
8   0.007812  2.164062  2.156250  0.148272   male       0     1  
9   0.015625  4.695312  4.679688  0.089920   male       0     1  
10  0.007812  2.812500  2.804688  0.200000   male       0     1  
11  0.007812  2.718750  2.710938  0.132351   male       0     1  

[12 rows x 23 columns]

The dummy variables are added to the columns.. So let's check the matrix to make our selection. 

In [6]:
df.corr()

meanfreq        sd    median       Q25       Q75       IQR  \
meanfreq  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
sd       -0.739039  1.000000 -0.562603 -0.846931 -0.161076  0.874660   
median    0.925445 -0.562603  1.000000  0.774922  0.731849 -0.477352   
Q25       0.911416 -0.846931  0.774922  1.000000  0.477140 -0.874189   
Q75       0.740997 -0.161076  0.731849  0.477140  1.000000  0.009636   
IQR      -0.627605  0.874660 -0.477352 -0.874189  0.009636  1.000000   
skew     -0.322327  0.314597 -0.257407 -0.319475 -0.206339  0.249497   
kurt     -0.316036  0.346241 -0.243382 -0.350182 -0.148881  0.316185   
sp.ent   -0.601203  0.716620 -0.502005 -0.648126 -0.174905  0.640813   
sfm      -0.784332  0.838086 -0.661690 -0.766875 -0.378198  0.663601   
mode      0.687715 -0.529150  0.677433  0.591277  0.486857 -0.403764   
centroid  1.000000 -0.739039  0.925445  0.911416  0.740997 -0.627605   
meanfun   0.460844 -0.466281  0.414909  0.545035  0.155091 -0.534462   
minfun    0.383937 -0.345609  0.337602  0.320994  0.258002 -0.222680   
maxfun    0.274004 -0.129662  0.251328  0.199841  0.285584 -0.069588   
meandom   0.536666 -0.482726  0.455943  0.467403  0.359181 -0.333362   
mindom    0.229261 -0.357667  0.191169  0.302255 -0.023750 -0.357037   
maxdom    0.519528 -0.482278  0.438919  0.459683  0.335114 -0.337877   
dfrange   0.515570 -0.475999  0.435621  0.454394  0.335648 -0.331563   
modindx  -0.216979  0.122660 -0.213298 -0.141377 -0.216475  0.041252   
female    0.337415 -0.479539  0.283919  0.511455 -0.066906 -0.618916   
male     -0.337415  0.479539 -0.283919 -0.511455  0.066906  0.618916   

              skew      kurt    sp.ent       sfm  ...   meanfun    minfun  \
meanfreq -0.322327 -0.316036 -0.601203 -0.784332  ...  0.460844  0.383937   
sd        0.314597  0.346241  0.716620  0.838086  ... -0.466281 -0.345609   
median   -0.257407 -0.243382 -0.502005 -0.661690  ...  0.414909  0.337602   
Q25      -0.319475 -0.350182 -0.648126 -0.766875  ...  0.545035  0.320994   
Q75      -0.206339 -0.148881 -0.174905 -0.378198  ...  0.155091  0.258002   
IQR       0.249497  0.316185  0.640813  0.663601  ... -0.534462 -0.222680   
skew      1.000000  0.977020 -0.195459  0.079694  ... -0.167668 -0.216954   
kurt      0.977020  1.000000 -0.127644  0.109884  ... -0.194560 -0.203201   
sp.ent   -0.195459 -0.127644  1.000000  0.866411  ... -0.513194 -0.305826   
sfm       0.079694  0.109884  0.866411  1.000000  ... -0.421066 -0.362100   
mode     -0.434859 -0.406722 -0.325298 -0.485913  ...  0.324771  0.385467   
centroid -0.322327 -0.316036 -0.601203 -0.784332  ...  0.460844  0.383937   
meanfun  -0.167668 -0.194560 -0.513194 -0.421066  ...  1.000000  0.339387   
minfun   -0.216954 -0.203201 -0.305826 -0.362100  ...  0.339387  1.000000   
maxfun   -0.080861 -0.045667 -0.120738 -0.192369  ...  0.311950  0.213987   
meandom  -0.336848 -0.303234 -0.293562 -0.428442  ...  0.270840  0.375979   
mindom   -0.061608 -0.103313 -0.294869 -0.289593  ...  0.162163  0.082015   
maxdom   -0.305651 -0.274500 -0.324253 -0.436649  ...  0.277982  0.317860   
dfrange  -0.304640 -0.272729 -0.319054 -0.431580  ...  0.275154  0.316486   
modindx  -0.169325 -0.205539  0.198074  0.211477  ... -0.054858  0.002042   
female   -0.036627 -0.087195 -0.490552 -0.357499  ...  0.833921  0.136692   
male      0.036627  0.087195  0.490552  0.357499  ... -0.833921 -0.136692   

            maxfun   meandom    mindom    maxdom   dfrange   modindx  \
meanfreq  0.274004  0.536666  0.229261  0.519528  0.515570 -0.216979   
sd       -0.129662 -0.482726 -0.357667 -0.482278 -0.475999  0.122660   
median    0.251328  0.455943  0.191169  0.438919  0.435621 -0.213298   
Q25       0.199841  0.467403  0.302255  0.459683  0.454394 -0.141377   
Q75       0.285584  0.359181 -0.023750  0.335114  0.335648 -0.216475   
IQR      -0.069588 -0.333362 -0.357037 -0.337877 -0.331563  0.041252   
skew     -0.080861 -0.336848 -0.061608 -0.305651 -0.304640 -0.169325   
kurt 

Our main interest is of course in the variables male and female. So if we take a look at the strongest correlation to these two variables, we can make our selection. Let's subset using: IQR, sp.ent, meanfun, sd, Q25, centroid, sfm, meanfreq en female.

In [7]:
df = df[['IQR','sp.ent', 'meanfun', 'sd', 'Q25','meanfreq','centroid','sfm', 'female']] #make this selection before you dropna
df = df.dropna() #first get rid of rows with empty cells 
df['female'].value_counts()

1    1584
0    1584
Name: female, dtype: int64

### Why am I using KNN?
I use the K nearest neighbor model. It determines the value of a datapoint by the ones nearby. It's explained by majority voting: by setting the parameter to e.g. 5, the classes of the 5 most close datapoints are checked. It thus will be predicted according to the majority. KNN is actually most of the times used for classification problems.


In [8]:
from sklearn.preprocessing import normalize #get the function needed to normalize our data.

X = df[['IQR','sp.ent','centroid', 'meanfun', 'sd', 'Q25','meanfreq','sfm']] #creating the X matrix
y = df['female'] #create the y-variable
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) #split the data, store it into different variables
X_train.head()

IQR    sp.ent  centroid   meanfun        sd       Q25  meanfreq  \
1866  0.174542  0.941751  0.141083  0.174179  0.088698  0.043235  0.141083   
767   0.100853  0.908148  0.167021  0.100181  0.057011  0.106603  0.167021   
2862  0.028535  0.813883  0.234486  0.179095  0.034413  0.223822  0.234486   
1064  0.105827  0.911062  0.197073  0.138659  0.057609  0.138016  0.197073   
270   0.108663  0.940602  0.142018  0.108932  0.071383  0.094235  0.142018   

           sfm  
1866  0.601127  
767   0.420416  
2862  0.166600  
1064  0.322959  
270   0.613985

Using the the build in model of sklearn, called KNeightborsClassifier we will start the prediction.

In [9]:
from sklearn.neighbors import KNeighborsClassifier #the object class we need

knn = KNeighborsClassifier(n_neighbors=3) #create a KNN-classifier with 3 neighbors
knn = knn.fit(X_train, y_train) #this fits the k-nearest neigbor model with the train data

### Let's evaluate our model!

In [10]:
knn.score(X_test, y_test) #calculating the fit (on the test data)

0.9758149316508938

The accuracy of our prediction is 97%. Let's however still have a look at other information such as precision, because there are other criteria we have to take into account deciding on making the best prediction.

In [11]:
from sklearn.metrics import confusion_matrix
y_test_pred = knn.predict(X_test) #the predicted values
cm = confusion_matrix(y_test, y_test_pred) #creates a "confusion matrix"
cm

array([[482,  12],
       [ 11, 446]])

Let's put it into a nice matrix for convenience: reads nicely.

In [12]:
conf_matrix = pd.DataFrame(cm, index=['Not female (actual)', 'Female (actual)'], columns = ['Not female (predicted)', 'Female (predicted)']) 
conf_matrix

Not female (predicted)  Female (predicted)
Not female (actual)                     482                  12
Female (actual)                          11                 446

These results show that 446 voice recordings were predicted to be female and actually were female. However 12 were predicted to be female but in fact were male.

### the recall and precision

$precision = \frac{446 }{ 446 + 12} = .97$
97% of predicted were actually female. So only 3% was predicted to be female but were in fact male.

$recall = \frac{446}{ 11 + 446} = .98$ 
98% of the female voices were correctly predicted to be female which means 2% is still missed. This is however a good recall.

### Future work

We can perhaps (for future work) improve the model by putting in even more different values for K